Check coverage of mutations

KtorZ opened the pull request:
Make sure that we do actually tests all mutations with enough
  coverage. This makes heavy use of generics to make it easy to maintain
  (i.e. add/remove mutations) and produces nice output for free:

  ```
  does not survive random adversarial mutations
    +++ OK, passed 800 tests.

    CloseMutation (800 in total):
    50.7% MutateParties
    17.4% MutateSnapshotNumberButNotSignature
    16.4% MutateSnapshotToIllFormedValue
    15.5% MutateSignatureButNotSnapshotNumber
  ```

KtorZ committed:
Check coverage of mutations

  Make sure that we do actually tests all mutations with enough
  coverage. This makes heavy use of generics to make it easy to maintain
  (i.e. add/remove mutations) and produces nice output for free:

  ```
  does not survive random adversarial mutations
    +++ OK, passed 800 tests.

    CloseMutation (800 in total):
    50.7% MutateParties
    17.4% MutateSnapshotNumberButNotSignature
    16.4% MutateSnapshotToIllFormedValue
    15.5% MutateSignatureButNotSnapshotNumber
  ```

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -223,11 +228,18 @@ healthySignature number = MultiSigned [sign sk snapshot | sk <- healthyPartyCred
   oneof
-    [ second ChangeHeadRedeemer <$> genChangeHeadRedeemer
-    , second ChangeHeadDatum <$> genChangeHeadDatum
+    [ bimap SomeMutationLabel ChangeHeadRedeemer <$> genChangeHeadRedeemer
```
Can we somehow combine the labels with the `Mutation`? It feels a bit doubled and/or maybe we can just have a simple mapping of `MutateSignatureButNotSnapshotNumber MockHead.Input -> ChangeHeadRedeemer MockHead.Input`? that way, it's a flat list of mutations and we can still leverage the enumerating labeling system you created?

KtorZ replied:
Possibly indeed, if we shove the `Mutation` inside the `SomeMutationLabel` (and rename the latter to `SomeMutation`) then it should work nicely.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/ContractSpec.hs:
```diff
@@ -125,13 +125,14 @@ prop_verifySnapshotSignatures =
       & applyMutation mutation
       & propTransactionDoesNotValidate
-      & withMaxSuccess 1000
+      & genericCoverTable @ 'UniformCoverage [lbl]
```
I don't understand what `[lbl]` is doing here?

ch1bo replied:
Is it just do indicate the `lbl` type as a proxy to `gCoverTableRequirements`?

KtorZ replied:
It counts the occurrence of that label in the property execution. A`Property` is intrinsically stateful and quickcheck records things during a property execution like label occurences in order to (a) determine whether the execution meets the coverage requirements and (b) provide a nice report at the end.

KtorZ replied:
I took some shortcut with `genericCoverTable` which is also calling `tabulate` and it's maybe the source of the confusion here. So, not only does this declare the table to cover, but it also record the occurrence of some of the cases we are after.

KtorZ replied:
Did this because I found the syntax `coverTable + tabulate` needlessly heavy. 99% of the time, they are called one after the other.

ch1bo commented on hydra-test-utils/src/Test/Hydra/Prelude.hs:
```diff
@@ -137,3 +156,83 @@ newtype ReasonablySized a = ReasonablySized a
+  where
+  gCoverTableRequirements =
+    let lengthI = toInteger . length
+        cs = dataTypeConstrs (dataTypeOf (Prelude.undefined :: t))
```
This seems to be the core idea of the generic mechanism. I ask myself whether we could use this via a simpler `uniformLabels :: Generic a => a -> [(String, Double)]` function and just pass this to `coverTable` or so?

KtorZ replied:
Hmmm. I think this question is the same as below but asked differently, so see answer on the other comment :no_mouth:

ch1bo commented on hydra-test-utils/src/Test/Hydra/Prelude.hs:
```diff
@@ -137,3 +156,83 @@ newtype ReasonablySized a = ReasonablySized a
+--
+--     forAll arbitrary $ \(lbl :: MyCoverLabel, arg) ->
+--         myProp arg
+--         & genericCoverTable @UniformCoverage [lbl]
```
Wouldn't we want to have `genericCoverTable arg` which of course requires `Generic` for the `arg` type, but it would be more concise IMO.

KtorZ replied:
Not exactly no. We still want / need some form of labels which are distinct from the arbitrary arg, otherwise we will end up labeling ALL generated values without any kind of grouping and this is generally not what we want.

KtorZ replied:
What we _could_ do is to only use the constructor name as a label (and not look at the fields) and make sure that `arg` is a sum-type which wrapped the label groups we care about. In this case, something like:

```hs
data CloseMutation
  = MutateSignatureButNotSnapshotNumber Mutation
  | MutateSnapshotNumberButNotSignature Mutation
  | MutateSnapshotToIllFormedValue Mutation
  | MutateParties Mutation
```

would work nicely.

KtorZ replied:
Okay, this should be fixed in 8ad9a06

ch1bo replied:
This looks a lot better now, yes. Although the documentation is outdated now :)

ch1bo reviewed (commented):
I love the idea, it feels a bit unpolished and too much generics type classes to me though. My comments are also a bit vague and I need to think abit more about it I guess. It looks like a good step forward though!

KtorZ committed:
Make 'genericCoverTable' work from argument and with only generic

  The Data constraint was a pain in the a** because it ripples down to all child types of a data-types and is therefore generally not applicable. Plus, by wrapping mutations inside a sum-type, and using the constructors as labels, it works nicely and removes some boilerplate from the consumer site.

ch1bo commented on hydra-prelude/src/Hydra/Prelude.hs:
```diff
@@ -26,6 +28,8 @@ module Hydra.Prelude (
   shrinkListAggressively,
   padLeft,
+  (^.),
+  (.~),
```
This seems to break an import in local-cluster -> CI red.

ch1bo commented on hydra-test-utils/src/Test/Hydra/Prelude.hs:
```diff
@@ -137,3 +155,113 @@ newtype ReasonablySized a = ReasonablySized a
+genericCoverTable xs =
+  coverTable tableName requirements . tabulate tableName (gLabelName . from <$> xs)
+ where
+  (tableName, requirements) = gCoverTableRequirements @distr (Proxy @(Rep a))
```
I still don't get why we would want to take `[a]` as we seemingly always only pass a single `[mutation]`? The documentation of [`tabulate`](https://hackage.haskell.org/package/QuickCheck-2.14.2/docs/Test-QuickCheck.html#v:tabulate) then states it is preferred over `label` if there is a varying amount of values to label.. but we have only one, so a single `label` should fit our purpose?

Furthermore, the `tableName` is always `""` in the current implementation of `gCoverTableRequirements` and parameterizing the distribution is YAGNI in my opinion.

If we could reduce the "noise" of generic type classes and instances down to a single `constructorNames :: Generic a => Proxy a -> [String]` label producing function, that would be preferred to me -> keep it simple. WDYT?

KtorZ replied:
> why we would want to take [a] as we seemingly always only pass a single [mutation]

For the moment indeed. Maybe YAGNI, but we had initially in mind to run _several mutations_ at once and combine them arbitrarily. Also, since this function is quite generic in the end (may be used for other test), I preferred keeping the API flexible instead of being needlessly restrictive. It works well for list, and it's also straightforward for the caller to wrap a single element in a singleton list.

> the tableName is always "" in the current implementation of gCoverTableRequirements

No! The tableName is set to the name of the overall wrapping type (here, `CloseMutation`). But the generic implementation is maybe a big hard to read. It's really a top-down structure with the `D1 _ _` constructor at the top; which is thus where the tableName is actually set. At lower levels, we collect other information but use a default, empty `""` table name (which is filled by the `D1 _ _` level).  

> and parameterizing the distribution is YAGNI in my opinion.

Arguably. It's more for the readability and for the sake o writing something a bit generic. It's not like it costs much effort to be frank :smile: 

> If we could reduce the "noise" of generic type classes and instances down to a single constructorNames :: Generic a => Proxy a -> [String]

Then I am curious as for how you'd write that function implementation without type classes :blush: ...

ch1bo reviewed (approved):
Only the CI prevents this from being merged now. I still want to try a more "boring Haskell" approach, but that does not prevent this from being merged - it's good stuff, anyways!

KtorZ committed:
Remove 'CoverageDistribution' kind for 'GCoverTable' + some documentation tweaks.

  Also introduced a type-family for the 'GCoverTableRequirements' class to avoid passing empty strings and default values when constructing the weight requirements generically. Instead, the return type of the function is inferred from the outer-most functor that wraps the generic representation. That is, fow lower levels where we simply look for accumulating constructors names, we use [String] whereas the final level (D1 c f) returns the table name and the labelled weights (String, [(String, Double)]). This is slightly more complicated but hopefully less confusing for a reader stumbling across this implementation.

KtorZ committed:
Fix conflict in prelude with Relude.Extra.Lens.
